doc/howto/rule-generation.rst

```diff
@@ -0,0 +1,266 @@
+Using Rule Generation
+=====================
+
+Sometimes it can be useful to generate Dune rules that depend on the file system
+layout or on the content of configuration files. This often happens for
+integration tests.
+
+In this document, we will see two ways to encode this behavior.
+
+We suppose that we are testing an executable named ``tool``. There are some
+input files named ``*.input``, output files named ``*.output``, and we want to
+ensure that when running ``tool`` on ``x.input``, the standard output
+corresponds to ``x.output``.
+
+The Generate-Include-Commit Pattern
+-----------------------------------
+
+.. note::
+
+   This is the most common way to do this. It has a couple drawbacks listed
+   below, but you should start with this pattern.
+
+What we are going to do is:
+
+- generate a ``dune.inc`` file;
+- include it in our main ``dune`` file;
+- commit the generated code in the source repository.
+
+This creates a loop: a program (the "generator") looks at the file system and
+creates a ``dune.inc`` file. Changes to the file system (for example, if a test
+is added) mean that a change in ``dune.inc`` will be promoted. These generated
+rules are included in the main ``dune`` file, so ``dune runtest`` will run the
+tests. Finally, the generated file is part of the source repository, so it is
+not necessary to run several commands to run the test suite.
+
+Let's expand a bit on how to achieve this.
+
+Generating a ``dune.inc`` File
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Create a ``gen`` subdirectory and create a ``gen.ml`` file in it:
+
+``gen/gen.ml``
+  .. code:: ocaml
+
+    let generate_rules base =
+      Printf.printf
+        {|
+            (rule
+             (with-stdout-to %s.gen
+              (run %%{bin:tool} %s.input)))
+        
+            (rule
+             (alias runtest)
+             (action
+              (diff %s.output %s.gen)))
+             |}
+        base base base base
+
+    let () =
+      Sys.readdir "."
+      |> Array.to_list
+      |> List.sort String.compare
+      |> List.filter_map (Filename.chop_suffix_opt ~suffix:".input")
+      |> List.iter generate_rules
+
+Create a ``dune`` file in that directory:
+
+``gen/dune``
+  .. code:: dune
+
+    (executable
+     (name gen))
+
+This defines an executable that lists ``*.input`` files in the current
+directory and outputs rules on its standard output.
+
+.. note::
+
+   It is important to sort the input files to ensure that the output is
+   independent on the order the files are returned by ``Sys.readdir``.
```

Comment:
```suggestion
   It is important to sort the input files to ensure that the output is
   independent from the order ``Sys.readdir`` returns the files .
```

Reply:
Don't we need to say
> from the order *in which* `Sys.readdir` returns the files?

(the context is that `Sys.readdir` returns a list of files, but in a seemingly random order)

Reply:
Good call! I updated the suggestion.